toolkit/tools/internal/ccachemanager/ccachemanager.go

```diff
@@ -0,0 +1,697 @@
+	TagFile   *CCacheArchive
+}
+
+type CCacheManager struct {
+	ConfigFileName    string
+	Configuration     *CCacheConfiguration
+	RootWorkDir       string
+	LocalDownloadsDir string
+	LocalUploadsDir   string
+	CurrentPkgGroup   *CCachePkgGroup
+	AzureBlobStorage  *azureblobstoragepkg.AzureBlobStorage
+}
+
+func buildRemotePath(arch, folder, name, suffix string) (string) {
+	return arch + "/" + folder + "/" + name + suffix
+}
+
+func (g *CCachePkgGroup) buildTarRemotePath(folder string) (string) {
+	return buildRemotePath(g.Arch, folder, g.Name, CCacheTarSuffix)
+}
+
+func (g *CCachePkgGroup) buildTagRemotePath(folder string) (string) {
+	return buildRemotePath(g.Arch, folder, g.Name, CCacheTagSuffix)
+}
+
+func (g *CCachePkgGroup) UpdateTagsPaths(remoteStoreConfig *RemoteStoreConfig, localDownloadsDir string, localUploadsDir string) {
+
+	tagFile := &CCacheArchive{
+		LocalSourcePath : localDownloadsDir + "/" + g.Name + CCacheTagSuffix,
+		RemoteSourcePath : g.buildTagRemotePath(remoteStoreConfig.TagsFolder),
+		LocalTargetPath : localUploadsDir + "/" + g.Name + CCacheTagSuffix,
+		RemoteTargetPath : g.buildTagRemotePath(remoteStoreConfig.TagsFolder),
+	}
+
+	logger.Log.Infof("  tag local source  : (%s)", tagFile.LocalSourcePath)
+	logger.Log.Infof("  tag remote source : (%s)", tagFile.RemoteSourcePath)
+	logger.Log.Infof("  tag local target  : (%s)", tagFile.LocalTargetPath)
+	logger.Log.Infof("  tag remote target : (%s)", tagFile.RemoteTargetPath)
+
+	g.TagFile = tagFile
+}
+
+func (g *CCachePkgGroup) UpdateTarPaths(remoteStoreConfig *RemoteStoreConfig, localDownloadsDir string, localUploadsDir string) {
+
+	tarFile := &CCacheArchive{
+		LocalSourcePath : localDownloadsDir + "/" + g.Name + CCacheTarSuffix,
+		RemoteSourcePath : g.buildTarRemotePath(remoteStoreConfig.DownloadFolder),
+		LocalTargetPath : localUploadsDir + "/" + g.Name + CCacheTarSuffix,
+		RemoteTargetPath : g.buildTarRemotePath(remoteStoreConfig.UploadFolder),
+	}
+
+	logger.Log.Infof("  tar local source  : (%s)", tarFile.LocalSourcePath)
+	logger.Log.Infof("  tar remote source : (%s)", tarFile.RemoteSourcePath)
+	logger.Log.Infof("  tar local target  : (%s)", tarFile.LocalTargetPath)
+	logger.Log.Infof("  tar remote target : (%s)", tarFile.RemoteTargetPath)
+
+	g.TarFile = tarFile
+}
+
+func (g *CCachePkgGroup) getLatestTag(azureBlobStorage *azureblobstoragepkg.AzureBlobStorage, containerName string) (string, error) {
+
+	logger.Log.Infof("  checking if (%s) already exists...", g.TagFile.LocalSourcePath)
+	_, err := os.Stat(g.TagFile.LocalSourcePath)
+	if err != nil {
+		// If file is not available locally, try downloading it...
+		logger.Log.Infof("  downloading (%s) to (%s)...", g.TagFile.RemoteSourcePath, g.TagFile.LocalSourcePath)
+		err = azureBlobStorage.Download(context.Background(), containerName, g.TagFile.RemoteSourcePath, g.TagFile.LocalSourcePath)
+		if err != nil {
+			logger.Log.Warnf("  unable to download ccache tag file.")
+			return "", err
```

Comment:
We are trying to transition to wrapped errors while minimizing logging as we update code. This would be:
```suggestion
			err = fmt.Errorf("unable to download ccache tag file:\n%w",err)
			return "", err
```

Reply:
If the warning logs still make sense, feel free to ignore this.

Reply:
None of ccache failures should indicate a problem with the build or the generated bits. They all just means that an optimization for build time failed to take place. So, I think using a warning for all of them will make them less distracting for people looking for actual blocking failures.
The reason I have remove the printing of the error is because the error has already been printed out by then. I initially printed the error ever where - and that results in seeing the same error 4 times in the logs (as it travels the stack). This made it very hard to know how many errors of that particular type occured (not readable).